In [11]:
require(tidyverse)
require(feather)
require(stringdist)

data_folder_path <- "C:\\Users\\javier\\WorkSpace"


# Load ICD10fi and ICD10who

In [56]:
# load ICD10fi
thl_icd10fi <- read_feather(file.path(data_folder_path,
                                      "ICD10fi",
                                      "THL_ICD10fi.feather") )

In [57]:
# load concept table and get the ICD10who
concept <- read_feather(file.path(data_folder_path,
                                      "OMOP_vocabulary_v5",
                                      "CONCEPT.feather") )
concept_icd10 <- concept  %>% filter(vocabulary_id == "ICD10")


# Study codes

**Summary ICD10fi coding marks**

In the Finnish version of the ICD10 there are two types of additional codes. 

- Classification codes: decrive ranges of codes other than the conventional ICD10 herarchy
    - `Code1-Code2` : from Code1 to Code2
- Reason codes: combine codes to add more info on what caused the diagnose, there are 4 marks
   - `Code1*Code2` : ”Oirekoodi”, Code2 indicates an additional symtom 
   - `Code1+Code2` : ”Syykoodi”, Code2 indicates the reason for Code1
   - `Code1#Code2` : ATC-koodi, Code2 is and ATC code indicating the medicine that caused Code1
   - `Code1&Code2` : ”Kasvainkoodi”, Code2 is and endocrinological disorder code that caused Code1


### How many are  clasification codes??

In [58]:
thl_icd10fi  %>% nrow

[1] 14681

In [129]:
thl_icd10fi  %>% distinct(CodeId)  %>% nrow

[1] 14681

In [59]:
thl_icd10fi_classif <- thl_icd10fi  %>% filter( grepl("-", CodeId))

In [60]:
thl_icd10fi_classif  %>% nrow

[1] 298

### How many are a direct match ??

In [61]:
# match ICD10fi codes to the ICD10who codes 
thl_icd10fi_match_icd10who <- inner_join(thl_icd10fi,
                concept_icd10 %>% rename(CodeId = concept_code),
                by = "CodeId")


In [62]:
thl_icd10fi_match_icd10who %>% nrow

[1] 9587

#### Do these agree in the definitio ??

In [63]:
# Count the differes in the number of charactes between the english names b
thl_icd10fi_match_icd10who  %>%  select(CodeId, English_name, concept_name, ShortName) %>% 
        mutate(same_name = stringdist(English_name,concept_name) )  %>% 
        count(same_name, sort = T) #%>% head(10)

same_name n   
1    0        8463
2   NA         226
3    2         150
4    3         149
5   20         114
6    4          91
7    1          84
8    7          32
9   17          26
10  10          20
11   5          18
12  14          14
13  37          13
14  42          13
15  15          12
16  19          12
17  39          12
18  22          11
19  45          11
20  59          10
21  18           8
22   8           7
23  32           6
24   6           5
25  16           5
26  27           5
27   9           4
28  11           4
29  12           4
30  24           4
... ...       ... 
32  13        3   
33  23        3   
34  28        3   
35  29        3   
36  38        3   
37  40        3   
38  43        3   
39  21        2   
40  25        2   
41  33        2   
42  36        2   
43  44        2   
44  47        2   
45  30        1   
46  35        1   
47  41        1   
48  46        1   
49  48        1   
50  49        1   
51  52        1   
52  55        1   
53  56        1   
54  57        1   
55  58        1   
56  61        1   
57  64        1   
58  65        1   
59  74        1   
60  76        1   
61  77        1

Most 8463 match the name exatly, 226 NA for these with no english name

#### Why the large differences ??

In [64]:
# Count the differes in the number of charactes between the english names b
thl_icd10fi_match_icd10who  %>%  select(CodeId, English_name, concept_name, ShortName) %>% 
        mutate(same_name = stringdist(English_name,concept_name) )  %>% 
        rename(ICD10fi_name = English_name, ICD10who_name = concept_name) %>% 
        filter(same_name>21)

CodeId
1   A09   
2   A49.1 
3   A74.0 
4   B96.5 
5   C81.1 
6   C81.2 
7   C81.3 
8   C82   
9   C83.6 
10  C83.9 
11  C84   
12  C84.4 
13  C90.2 
14  C91.5 
15  C92.1 
16  C92.2 
17  C96.0 
18  D39.1 
19  D47.2 
20  D76   
21  E11.1 
22  E11.5 
23  E11.6 
24  E11.7 
25  E11.8 
26  E11.9 
27  F66.8 
28  F79.9 
29  G23.1 
30  G23.2 
... ...   
105 R89.5 
106 R89.6 
107 R89.7 
108 R89.8 
109 R89.9 
110 T81.0 
111 U06.9 
112 U07.1 
113 W01   
114 W31   
115 W68   
116 V10.2 
117 V10.9 
118 V11   
119 V11.2 
120 V11.9 
121 V14.2 
122 V14.9 
123 V16.2 
124 V16.9 
125 V28.2 
126 V28.9 
127 V48.2 
128 V48.9 
129 V87   
130 V88   
131 Y57   
132 Y58   
133 Y59   
134 Z74.0 
    ICD10fi_name                                                                                             
1   Diarrhoea and gastroenteritis of presumed infectious origin                                              
2   Streptococcal infection, unspecified                                                                     
3   Paratrachoma                                                                                             
4   Pseudomonas (aeruginosa)(mallei)(pseudomallei) as the cause of diseases classified to other chapters     
5   Nodular sclerosis                                                                                        
6   Mixed cellularity                                                                                        
7   Lymphocytic depletion                                                                                    
8   Follicular [nodular] non-Hodgkin's lymphoma                                                              
9   Undifferentiated (diffuse)                                                                               
10  Diffuse non-Hodgkin's lymphoma, unspecified                                                              
11  Peripheral and cutaneous T-cell lymphomas                                                                
12  Peripheral T-cell lymphoma                                                                               
13  Plasmacytoma, extramedullary                                                                             
14  Adult T-cell leukaemia                                                                                   
15  Chronic myeloid leukaemia                                                                                
16  Subacute myeloid leukaemia                                                                               
17  Letterer-Siwe disease                                                                                    
18  Ovary                                                                                                    
19  Monoclonal gammopathy                                                                                    
20  Certain diseases involving lymphoreticular tissue and reticulohistiocytic system                         
21  Non-insulin-dependent diabetes mellitus - With ketoacidosis                                              
22  Non-insulin-dependent diabetes mellitus - With peripheral circulatory complications                      
23  Non-insulin-dependent diabetes mellitus - With other specified complications                             
24  Non-insulin-dependent diabetes mellitus - With multiple complications                                    
25  Non-insulin-dependent diabetes mellitus - With unspecified complications                                 
26  Non-insulin-dependent diabetes mellitus - Without complications                                          
27  Other specified disorders of psychosexual development                                                    
28  Unspecified mental retardation                                                                           
29  Progressive supranuclear ophthalmoplegia                                                                 
30  Striatonigral dege

Seems the concept are the same, descrived in different way. 

### How many more match if only the diagnose code is taken? 

In [85]:
# match ICD10fi's diagnose codes to the ICD10who codes 
matched_ids <- c(thl_icd10fi_classif$CodeId, thl_icd10fi_match_icd10who$CodeId)

thl_icd10fi_code1_match_icd10who <- inner_join( thl_icd10fi  %>% filter(!(CodeId %in% matched_ids)) , 
                concept_icd10 %>% rename(Code1 = concept_code),
                by = "Code1")


In [87]:
thl_icd10fi_code1_match_icd10who  %>% nrow

[1] 2105

### How many ICD10fi diagnose codes do not exist in ICD10who? 

In [88]:
matched_ids <- c(thl_icd10fi_classif$CodeId, thl_icd10fi_match_icd10who$CodeId, thl_icd10fi_code1_match_icd10who$CodeId)

thl_icd10fi_new <- thl_icd10fi  %>% filter(!(CodeId %in% matched_ids)) 

In [91]:
# which not composed codes are new
thl_icd10fi_new %>%
select(CodeId, English_name, ShortName)

CodeId      
1    A28.10      
2    A28.11      
3    A28.19      
4    A31.80      
5    A31.81      
6    A31.84      
7    A31.89      
8    A50.50      
9    A50.51      
10   A50.52      
11   A50.54      
12   A50.55      
13   A50.59      
14   A50.59+I98.0
15   A50.59+M03.1
16   A50.59+M90.2
17   A52.70      
18   A52.71      
19   A52.72      
20   A52.73      
21   A52.74      
22   A52.75      
23   A52.79      
24   A52.79+H19.0
25   A52.79+H32.0
26   A52.79+H58.8
27   A52.79+J99.8
28   A52.79+K67.2
29   A52.79+K77.0
30   A52.79+L99.8
...  ...         
2662 ZA1.0       
2663 ZA2.0       
2664 ZA3.0       
2665 ZA4.0       
2666 ZA5.0       
2667 ZA5.1       
2668 ZA6.0       
2669 ZA7.0       
2670 ZA8.0       
2671 ZA8.1       
2672 ZA8.2       
2673 ZA8.3       
2674 ZA8.4       
2675 ZB0.0       
2676 ZB0.1       
2677 ZB0.2       
2678 ZB1.0       
2679 ZB1.1       
2680 Z03.80      
2681 Z03.89      
2682 Z13.80      
2683 Z13.81      
2684 Z13.82      
2685 Z13.88      
2686 Z22.30      
2687 Z22.38      
2688 Z51.80      
2689 Z51.88      
2690 Z52.80      
2691 Z52.89      
     English_name                                                            
1    NA                                                                      
2    NA                                                                      
3    NA                                                                      
4    NA                                                                      
5    NA                                                                      
6    NA                                                                      
7    Other mycobacterial infections                                          
8    NA                                                                      
9    NA                                                                      
10   NA                                                                      
11   NA                                                                      
12   NA                                                                      
13   Other late congenital syphilis, symptomatic                             
14   Late congenital cardiovascular syphilis+                                
15   Clutton's joints+                                                       
16   Late congenital syphilitic: osteochondropathy+                          
17   NA                                                                      
18   NA                                                                      
19   NA                                                                      
20   NA                                                                      
21   NA                                                                      
22   NA                                                                      
23   Other symptomatic late syphilis                                         
24   Late syphilitic episcleritis+                                           
25   Late syphilitic chorioretinitis+                                        
26   Late syphilitic oculopathy NEC+                                         
27   Syphilis [stage unspecified] of lung+                                   
28   Syphilitic peritonitis                                                  
29   Syphilitic liver disease                                                
30   Late syphilitic leukoderma+                                             
...  ...                                                                     
2662 Shoulder region                                                         
2663 Upper arm                                                               
2664 Forearm                                                                 
2665 Hand                                                                    
2666 Pelvic region                                                           
266

In [92]:
#which are new in the level higher than 6
thl_icd10fi_new  %>% 
select(CodeId, English_name, ShortName) %>% 
mutate(group = nchar(CodeId) #str_sub(concept_code, 0, 1)
      ) %>% filter(group<6)

CodeId
1   B07.9 
2   F61.0 
3   F61.1 
4   W71   
5   X85.0 
6   X85.2 
7   X85.8 
8   X85.9 
9   X90.0 
10  X90.2 
11  X90.8 
12  X90.9 
13  X91.0 
14  X91.2 
15  X91.8 
16  X91.9 
17  X92.0 
18  X92.2 
19  X92.8 
20  X92.9 
21  X93.0 
22  X93.2 
23  X93.8 
24  X93.9 
25  X94.0 
26  X94.2 
27  X94.8 
28  X94.9 
29  X95.0 
30  X95.2 
... ...   
54  Y94.0 
55  Y94.1 
56  Y94.2 
57  Y94.4 
58  Y94.5 
59  Y94.8 
60  Y95.0 
61  Y96.0 
62  Y96.2 
63  Y96.8 
64  Y96.9 
65  ZA0.0 
66  ZA1.0 
67  ZA2.0 
68  ZA3.0 
69  ZA4.0 
70  ZA5.0 
71  ZA5.1 
72  ZA6.0 
73  ZA7.0 
74  ZA8.0 
75  ZA8.1 
76  ZA8.2 
77  ZA8.3 
78  ZA8.4 
79  ZB0.0 
80  ZB0.1 
81  ZB0.2 
82  ZB1.0 
83  ZB1.1 
    English_name                                                                 
1   Verruca simplex                                                              
2   Mixed personality disorders                                                  
3   NA                                                                           
4   Drowning to ice                                                              
5   Assault by drugs, medicaments and biological substances by spouse or partner 
6   Assault by drugs, medicaments and biological substances by friend            
7   Assault by drugs, medicaments and biological substances by other known person
8   Assault by drugs, medicaments and biological substances by unknown person    
9   Assault by unspecified chemical or noxious substance by spouse or partner    
10  Assault by unspecified chemical or noxious substance by friend               
11  Assault by unspecified chemical or noxious substance by other known person   
12  Assault by unspecified chemical or noxious substance by unknown person       
13  Assault by hanging, strangulation and suffocation by spouse or partner       
14  Assault by hanging, strangulation and suffocation by friend                  
15  Assault by hanging, strangulation and suffocation by other known person      
16  Assault by hanging, strangulation and suffocation by unknown person          
17  Assault by drowning and submersion by spouse or partner                      
18  Assault by drowning and submersion by friend                                 
19  Assault by drowning and submersion other specified person                    
20  Assault by drowning and submersion by unknown person                         
21  NA                                                                           
22  NA                                                                           
23  NA                                                                           
24  NA                                                                           
25  NA                                                                           
26  NA                                                                           
27  NA                                                                           
28  NA                                                                           
29  Assault by other and unspecified firearm discharge by spouse or partner      
30  Assault by other and unspecified firearm discharge by friend                 
... ...                                                                          
54  Accident at home                                                             
55  NA                                                                           
56  Accident with sport                                                          
57  NA                                                                           
58  NA                                                                           
59  Other leisure activity accident                                              
60  Accident in a hospital or hospital associated external cause                 
61  Work-related condition                                                       
62  School or kindergarten related conditi

In [94]:
# what groups have more new ones 
thl_icd10fi_new  %>% 
select(CodeId, English_name, ShortName) %>% 
mutate(group = str_sub(CodeId, 0, 1)) %>% count(group, sort = T)

group n  
1  C     770
2  F     497
3  Q     348
4  K     313
5  D     130
6  I     105
7  E      89
8  G      62
9  H      50
10 X      48
11 A      39
12 Y      36
13 B      35
14 S      32
15 Z      31
16 P      26
17 N      16
18 O      16
19 L      15
20 J      10
21 R      10
22 T       7
23 M       3
24 V       2
25 W       1

# Proposed matching  

1. Ignore the new ICD10fi clasification codes, these are not supose to be used as diagnose
2. Match ICD10fi to ICD10who only based on the diagnose code `code1`
3. These new ICD10fi that dont exist in ICD10who, match to the parent code

In [122]:
# ICD10fi code mathches the ICD10who
thl_icd10fi_clas <- thl_icd10fi  %>% 
    filter( CodeId %in% concept_icd10$concept_code)  %>% 
    mutate( ICD10who = CodeId , ICD10who_match_level = 0 )

In [150]:
# match diagnose codes to the upper level code
thl_icd10fi_new_1 <- inner_join( thl_icd10fi_new  %>% 
                                # remove the last digit of the code 
                                mutate(ICD10who = str_sub(Code1, 0, -2))%>% 
                                # remove the last digit of the code if it is a point
                                mutate(ICD10who = sub("\\.$", "", ICD10who)) , 
                concept_icd10 %>% rename(ICD10who = concept_code),
                by = "ICD10who")


In [151]:
thl_icd10fi_new_1  %>%  nrow

[1] 2587

In [161]:
# match diagnose codes to the upper level code
thl_icd10fi_new_2 <- left_join( thl_icd10fi_new  %>% filter( !(CodeId %in% thl_icd10fi_new_1$CodeId)) %>%  
                                # remove the last digit of the code 
                                mutate(ICD10who = str_sub(Code1, 0, -3))%>% 
                                # remove the last digit of the code if it is a point
                                mutate(ICD10who = sub("\\.$", "", ICD10who)) , 
                concept_icd10 %>% rename(ICD10who = concept_code),
                by = "ICD10who")

In [162]:
thl_icd10fi_new_2 %>%  nrow

[1] 104

In [173]:
# Join all 
thl_icd10fi_matched <- bind_rows(
    # classifiction codes
    thl_icd10fi  %>% filter( CodeId %in%thl_icd10fi_classif$CodeId)  %>% 
    mutate(ICD10who = as.character(NA), ICD10who_match_level = "classification" ), 
    # perfect matches
    thl_icd10fi  %>% filter( CodeId %in%thl_icd10fi_match_icd10who$CodeId)  %>% 
    mutate(ICD10who = CodeId, ICD10who_match_level = "full_match" ), 
    # matched with the diagnose 
    thl_icd10fi  %>% filter( CodeId %in%thl_icd10fi_code1_match_icd10who$CodeId)  %>% 
    mutate(ICD10who = Code1, ICD10who_match_level = "diagnose_match" ),
    #
    thl_icd10fi_new_1 %>% select(CodeId:ICD10who)%>% 
    mutate(ICD10who_match_level = "diagnose_match_parent" ),
    #
    thl_icd10fi_new_2 %>% select(CodeId:ICD10who)%>% 
    mutate(ICD10who_match_level =  "diagnose_match_grandparent" )
)%>% arrange(CodeId)  %>% 
mutate(ICD10who_match_level = factor(ICD10who_match_level, 
                                     levels = c("classification", 
                                                "full_match", 
                                                "diagnose_match",
                                                "diagnose_match_parent", 
                                                "diagnose_match_grandparent")
                                    )
      )

In [174]:
thl_icd10fi_matched  %>% count(ICD10who_match_level)

ICD10who_match_level       n   
1 classification              298
2 full_match                 9587
3 diagnose_match             2105
4 diagnose_match_parent      2587
5 diagnose_match_grandparent  104

In [175]:
# load ICD10fi
write_feather(thl_icd10fi_matched, file.path(data_folder_path,
                                      "ICD10fi",
                                      "THL_ICD10fi_matched_ICD10who.feather") )